model-optimizer/extensions/ops/gathernd.py

```diff
@@ -0,0 +1,103 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import numpy as np
+
+from mo.front.common.partial_infer.utils import int64_array
+from mo.graph.graph import Node, Graph
+from mo.ops.op import Op
+
+
+class GatherND(Op):
+    op = 'GatherND'
+
+    def __init__(self, graph: Graph, attrs: dict):
+        mandatory_props = {
+            'kind': 'op',
+            'type': __class__.op,
+            'op': __class__.op,
+            'version': 'opset5',
+            'infer': __class__.infer,
+            'in_ports_count': 2,
+            'out_ports_count': 1,
+        }
+        super().__init__(graph, mandatory_props, attrs)
+
+    def backend_attrs(self):
+        return ['batch_dims']
+
+    @staticmethod
+    def infer(node: Node):
+        node_name = node.soft_get('name', node.id)
+        connected_in_ports = [port for port in node.in_ports().values() if not port.disconnected()]
+        assert len(connected_in_ports) == 2, \
+            "Incorrect number of inputs for {} node".format(node_name)
+
+        data_shape = node.in_port(0).data.get_shape()
+        data_value = node.in_port(0).data.get_value()
+        indices_shape = node.in_port(1).data.get_shape()
+        indices_value = node.in_port(1).data.get_value()
+
+        batch_dims = 0
+        if node.has_valid('batch_dims'):
+            batch_dims = node.batch_dims
+
+        # check that a number of batch dimensions is less than both ranks of data and indices tensors
+        assert batch_dims < len(data_shape), "Number of batch dimensions must be less than a rank of data"
+        assert batch_dims < len(indices_shape), "Number of batch dimensions must be less than a rank of indices"
+
+        # check that batch dimensions of data and indices are the same
+        for batch_dim in range(batch_dims):
+            assert data_shape[batch_dim] == indices_shape[batch_dim], \
+                "The dimension {} for data and indices tensors must be the same".format(batch_dim)
+
+        # check ranks of input tensors
+        assert len(data_shape) > 0, "Data must not be a scalar"
+        assert len(indices_shape) > 0, "Indices must not be a scalar"
+        assert (batch_dims + indices_shape[-1]) <= len(data_shape), \
+            "Length of a tuple with indices must not exceed a rank of data tensor excluding batch dimensions"
+
+        # compute output shape
+        number_batches = [np.prod(data_shape[:batch_dims]).tolist()] if batch_dims > 0 else list()
+        slice_shape = list(data_shape[(batch_dims + indices_shape[-1]):])
+        output_shape = number_batches + list(indices_shape[batch_dims:-1]) + slice_shape
+        node.out_port(0).data.set_shape(int64_array(output_shape))
+
+        # compute output value if all input values are defined
+        if data_value is not None and indices_value is not None:
```

Comment:
nit: move value propagation to a separate function just to make shape infer function a bit more clear.

Reply:
I did try but it looks worse from my point. The function requires `data_shape`, `indices_shape`, `output_shape` and these local variable I need to pass through arguments or create inside the function. This thing makes the function awkward.